core/src/smartcontracts/isi/world.rs

```diff
@@ -174,6 +174,66 @@ pub mod isi {
         }
     }
 
+    impl Execute for Grant<PermissionToken, Role> {
+        #[metrics(+"grant_role_permission")]
+        fn execute(self, _authority: &AccountId, wsv: &mut WorldStateView) -> Result<(), Error> {
+            let role_id = self.destination_id;
+            let permission_token = self.object;
+            let permission_token_id = permission_token.definition_id.clone();
+
+            if !wsv
+                .permission_token_schema()
+                .token_ids
+                .contains(&permission_token_id)
+            {
+                return Err(FindError::PermissionToken(permission_token_id).into());
+            }
+
+            let Some(role) = wsv.world.roles.get_mut(&role_id) else {
+                return Err(FindError::Role(role_id).into());
+            };
+
+            if !role.permissions.insert(permission_token.clone()) {
+                return Err(RepetitionError {
+                    instruction_type: InstructionType::Grant,
+                    id: permission_token.definition_id.into(),
+                }
+                .into());
+            }
+
+            wsv.emit_events(Some(RoleEvent::PermissionAdded(RolePermissionChanged {
+                role_id,
+                permission_token_id,
+            })));
+
+            Ok(())
+        }
+    }
+
+    impl Execute for Revoke<PermissionToken, Role> {
+        #[metrics(+"grant_role_permission")]
+        fn execute(self, _authority: &AccountId, wsv: &mut WorldStateView) -> Result<(), Error> {
+            let role_id = self.destination_id;
+            let permission_token = self.object;
+            let permission_token_id = permission_token.definition_id.clone();
+
+            let Some(role) = wsv.world.roles.get_mut(&role_id) else {
+                return Err(FindError::Role(role_id).into());
+            };
+
+            if !role.permissions.remove(&permission_token) {
```

Comment:
What do you mean by "prompting to gracefully unregister the role"?

Reply:
the question is should we allow a `Role` without a permission? If no, then we should make sure that it's constructor doesn't allow constructing permissionless role and also that burning/revoking a permission from a role is disallowed if it's the last one

Reply:
I think empty roles are ok, user might want to gradually add/remove permissions without removing role.
Also since removing role remove it from every account i want to be very careful with removing empty roles.

Reply:
> Also since removing role remove it from every account i want to be very careful with removing empty roles

to be precise, I don't think that we should remove empty roles. That's a horrible default, though this logic can be implemented in a custom executor by the user. I was only rethinking whether having empty roles should be allowed.

Reply:
I just meant it would be nice if there was a notification of empty roles.
I admit such a message encouraging unregister roles might be an extra care

Reply:
> I just meant it would be nice if there was a notification of empty roles.

Probably events could be used for this purpose. Let's create separate ticket for this.